File: 0994-rotting-oranges/0994-rotting-oranges.py
class Solution:
    def orangesRotting(self, grid: list[list[int]]) -> int:
        ROWS = len(grid)
        COLS = len(grid[0])
        DIRECTIONS = [ [0, -1], [0, 1], [1, 0], [-1, 0] ]

        fresh_oranges = 0
        queue = deque()
        
        for r in range(ROWS):
            for c in range(COLS):
                if grid[r][c] == 1:
                    fresh_oranges += 1
                elif grid[r][c] == 2:
                    queue.append((r, c))
        
        minutes = 0

        while queue:
            queue_length = len(queue)

            for _ in range(queue_length):
                r, c = queue.popleft()
                
                for dr, dc in DIRECTIONS:
                    nr = dr + r
                    nc = dc + c

                    if (
                        0 <= nr < ROWS
                        and 0 <= nc < COLS
                        and grid[nr][nc] == 1
                    ):
                        grid[nr][nc] = 2
                        fresh_oranges -= 1
                        queue.append((nr, nc))
            
            if len(queue):
                minutes += 1
        
        return minutes if fresh_oranges == 0 else -1

        

File: 0994-rotting-oranges/0994-rotting-oranges.js
/**
 * @param {number[][]} grid
 * @return {number}
 */
var orangesRotting = function(grid) {
    let rows = grid.length, cols = grid[0].length;
    let directions = [ [1, 0], [-1, 0], [0, 1], [0, -1] ];
    let queue = [];
    let minM = 0;
    let freshO = 0;
    
    for( let r=0; r < rows; r++ ) {
        for( let c=0; c<cols; c++ ) {
            if( grid[r][c] === 2 ) {
                queue.push([r, c]);
            } else if( grid[r][c] === 0 )
                grid[r][c] = 2
            else
                freshO++;
        }
    }
    
    if( freshO === 0 )
        return 0;
    
    while( queue.length > 0 ) {
        let length = queue.length;
        
        for( let i=0; i < length; i++ ) {
            let [r, c] = queue.shift();
            
            directions.map((direction) => {
                let [ rD, cD ] = direction;
                let rN = rD+r, cN = cD+c;
                
                if( rN >= 0 && cN >= 0 && rN< rows && 
                   cN < cols && grid[rN][cN] === 1 ) {
                    queue.push([rN, cN]);
                    grid[rN][cN] = 2;
                    freshO--;
                }
            });
        }
        
        if(queue.length > 0) {
            minM++;
        }
    }
    
    return freshO == 0 ? minM : -1;
};